pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/junit/JUnitTestsShouldIncludeAssertRule.java

```diff
@@ -29,36 +37,73 @@ public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
     @Override
     public Object visit(ASTMethodDeclaration method, Object data) {
         if (isJUnitMethod(method, data)) {
-            if (!containsAssert(method.getBlock(), false) && !containsExpect(method.jjtGetParent())) {
-                addViolation(data, method);
+            if (!isExpectAnnotated(method.jjtGetParent())) {
+                Scope classScope = method.getScope().getParent();
+                Map<String, List<NameOccurrence>> expectables = getRuleAnnotatedExpectedExceptions(classScope);
+                
+                if (!containsExpectOrAssert(method.getBlock(), expectables)) {
+                    addViolation(data, method);
+                }
             }
         }
         return data;
     }
 
-    private boolean containsAssert(Node n, boolean assertFound) {
-        if (!assertFound) {
-            if (n instanceof ASTStatementExpression) {
-                if (isAssertOrFailStatement((ASTStatementExpression) n)) {
-                    return true;
-                }
+    private boolean containsExpectOrAssert(Node n, Map<String, List<NameOccurrence>> expectables) {
+        if (n instanceof ASTStatementExpression) {
+            if (isExpectStatement((ASTStatementExpression) n, expectables)
+                    || isAssertOrFailStatement((ASTStatementExpression) n)) {
+                return true;
             }
-            if (!assertFound) {
-                for (int i = 0; i < n.jjtGetNumChildren() && !assertFound; i++) {
-                    Node c = n.jjtGetChild(i);
-                    if (containsAssert(c, assertFound)) {
-                        return true;
-                    }
+        } else {
+            for (int i = 0; i < n.jjtGetNumChildren(); i++) {
+                Node c = n.jjtGetChild(i);
+                if (containsExpectOrAssert(c, expectables)) {
+                    return true;
                 }
             }
         }
         return false;
     }
+    
+    /**
+     * Gets a list of NameDeclarations for all the fields that have type
+     * ExpectedException and have a Rule annotation.
+     *
+     * @param classScope
+     *            The class scope to search for
+     * @return See description
+     */
+    private Map<String, List<NameOccurrence>> getRuleAnnotatedExpectedExceptions(Scope classScope) {
+        Map<String, List<NameOccurrence>> result = new HashMap<>();
+        Map<NameDeclaration, List<NameOccurrence>> decls = classScope.getDeclarations();
+
+        for (NameDeclaration decl : decls.keySet()) {
+            Node parent = decl.getNode().jjtGetParent().jjtGetParent().jjtGetParent();
+            if (parent.hasDescendantOfType(ASTAnnotation.class)
+                    && parent.getFirstChildOfType(ASTFieldDeclaration.class) != null) {
+                if (!"Rule"
```

Comment:
bare in mind that, even if non-conventional, people may simply write `@org.junit.Rule`too